### DISCUSSION TITLE
C# readonly, property 관련 질문

### DISCUSSION THREAD

**Discussion by fkdl0048** [fkdl0048/Mentoring-Project#42] [Q&A] [answered]
빌더 패턴을 조금 알아보던 중에 c# 기능을 활용해서 만들어 볼려고 공부 중에 궁금한 점이 있어서 질문 남깁니다.  

1. readonly로 만든 필드와 get만 열어둔 프로퍼티의 차이점  

```cs
  private readonly int a;
  public int A {get;}
```

readonly의 경우 생성 단계에서만 값이 할당 가능하기 때문에 이후에도 필드로서 불변성을 가진다.
A의 경우 외부에서 열람만 가능하지만 내부 객체에선 값이 변경될 수 있다.   

이 정도 차이점만 가지는게 맞을까요?  

2. 람다와 프로퍼티의 차이점  

```cs
public string GetFont => font; // 프로퍼티
public string GetFont() => font; // 람다
```

동작이 동일하게 동작하는 것 같은데 차이점이나 장점이 있을까요?

**Comment by jongfeel** [comment, marked as answer]
1.

readonly의 경우는 맞습니다. 그런데 primitive 타입은 값이기 때문에 그럴 수 있지만 객체라면 참조이기 때문에 참조 값에 대한 불변성은 가져도 참조가 가리키는 값에 대해서는 불변이 아닙니다.

A의 경우 A property는 접근함에 있어서 기대값을 가지고 호출하는 것으로 public 인터페이스로 이해할 수는 있지만
내부에서 값이 변경될 수 있는지 없는지는 A를 포함하는 클래스의 책임에 달려 있습니다.

2.

동작은 동일해도 역시 기대하는 수준이 다를 수 있습니다.
Property의 경우는 속성의 값이라고 이해할 수 있으므로 별다른 연산 없이 값 자체를 얻을 수 있다는 기대를 할 수 있습니다.
Method의 경우는 호출해서 어떤 로직을 수행하고 그 결과를 얻는다는 기대를 할 수 있습니다. 즉 수행의 결과를 얻는다는 뜻이 더 강하므로 객체에게 뭔가 명령해서 수행시키는 느낌을 더 줄 수 있습니다.

코드를 짜는 사람이야 결과가 같으니 무슨 차이가 있는지 크게 신경 쓰지 않지만
이 코드를 사용하는 사람은 호출의 결과는 같을지라도 내부에서 뭔가 실행의 차이가 있을 거라고 기대할 수 있는 거죠.

**Reply by fkdl0048** [reply]
감사합니다.. 진짜 생각하지도 못했는데..  

1. 객체에 대한 참조는 책에서 나온 깊은 오용성이나 참조에 대한 무서움이 조금 생긴 것 같습니다.
2. 다른 개발자가 볼 때의 시선으로 생각해보지 못한 것 같습니다.. 감사합니다
